ApplicationInsightsInternalLogger/src/main/java/com/microsoft/applicationinsights/internal/logger/InternalLogger.java

```diff
@@ -85,33 +103,79 @@ private InternalLogger() {
      * @param loggerData The data for the internal logger
      */
     public synchronized void initialize(String loggerOutputType, Map<String, String> loggerData) {
-        if (!initialized) {
-            try {
-                String loggerLevel = loggerData.get(LOGGER_LEVEL);
-                if (Strings.isNullOrEmpty(loggerLevel)) {
-                    // The user didn't specify the logging level, therefore by default we set that to 'TRACE'
-                    loggingLevel = LoggingLevel.TRACE;
-                    setLoggerOutput(loggerOutputType, loggerData);
-                } else {
-                    try {
-                        // Try to match the user request logging level into our enum.
-                        loggingLevel = LoggingLevel.valueOf(loggerLevel.toUpperCase());
-                        setLoggerOutput(loggerOutputType, loggerData);
-                    } catch (Exception e) {
-                        onInitializationError(String.format("Error: Illegal value '%s' for the SDK internal logger. Logging level is therefore set to 'OFF'", loggerLevel));
-                    }
-                }
+        if (initialized) {
+            return;
+        }
 
-                final String utcId = "UTC";
+        try {
+            String loggerLevel = loggerData.get(LOGGER_LEVEL);
+            if (Strings.isNullOrEmpty(loggerLevel)) {
+                // The user didn't specify the logging level, therefore by default we set that to 'TRACE'
+                loggingLevel = LoggingLevel.TRACE;
+                setLoggerOutput(loggerOutputType, loggerData);
+            } else {
                 try {
-                    dateFormatter.setTimeZone(TimeZone.getTimeZone(utcId));
+                    // Try to match the user request logging level into our enum.
+                    loggingLevel = LoggingLevel.valueOf(loggerLevel.toUpperCase());
+                    setLoggerOutput(loggerOutputType, loggerData);
                 } catch (Exception e) {
-                    new ConsoleLoggerOutput().log(String.format("Failed to find timezone with id='%s'. Using default '%s'", utcId, dateFormatter.getTimeZone().getDisplayName()));
+                    onInitializationError(String.format("Error: Illegal value '%s' for the SDK internal logger. Logging level is therefore set to 'OFF'", loggerLevel));
                 }
-            } finally {
-                initialized = true;
+            }
+
+            final String utcId = "UTC";
+            try {
+                dateFormatter.setTimeZone(TimeZone.getTimeZone(utcId));
+            } catch (Exception e) {
+                new ConsoleLoggerOutput().log(String.format("Failed to find timezone with id='%s'. Using default '%s'", utcId, dateFormatter.getTimeZone().getDisplayName()));
+            }
+        } finally {
+            initialized = true;
+        }
+    }
+
+    public synchronized boolean systemPropertyInitialize() {
```

Comment:
This `synchronized` block is a costly lock operation especially when checking on every log message. It will significantly reduce the performance of SDK when trace logging is turned on`.

Also, I am not sure why a synchronized is needed here. Are there any known scenarios where multiple threads can try to simultaneously initialize this. Atleast not that I know of.

Reply:
Do you envision any scenario for thread safety? Do you see any cases where there can be two threads do initialization at same time. In my mind there are none. Please educate me with one, than I am fine to keep this sync. Otherwise its just wise to remove it imho.

Reply:
We can try to determine if thread safety is necessary in a different project. My intention was not to change how this class works, just to add a feature.